Source/OCMock/OCMNonRetainingObjectReturnValueProvider.m

```diff
@@ -33,12 +33,18 @@ - (void)handleInvocation:(NSInvocation *)anInvocation
     {
         @throw [NSException exceptionWithName:NSInvalidArgumentException reason:@"Expected invocation with object return type. Did you mean to use andReturnValue: instead?" userInfo:nil];
     }
-    NSString *sel = NSStringFromSelector([anInvocation selector]);
-    if([sel hasPrefix:@"alloc"] || [sel hasPrefix:@"new"] || [sel hasPrefix:@"copy"] || [sel hasPrefix:@"mutableCopy"])
+
+    if(OCMIsInvocationAllocFamily(anInvocation) || OCMIsInvocationNewFamily(anInvocation) || OCMIsInvocationCopyFamily(anInvocation) || OCMIsInvocationMutableCopyFamily(anInvocation))
     {
         // methods that "create" an object return it with an extra retain count
         [returnValue retain];
     }
+    else if(OCMIsInvocationInitFamily(anInvocation))
+    {
+        // init family methods "consume" self and retain their return value. Do the retain first in case the return value and self are the same.
+        [returnValue retain];
```

Comment:
I don't think I understand why the value should be retained. In an `alloc`/`init` chain the retain count is modified only by `alloc`, and not by `init`, right?

Reply:
ARC expects init family methods to consume "self" and retain their return value (see https://clang.llvm.org/docs/AutomaticReferenceCounting.html#method-families). So in a standard "ARC" call flow where you have 

`foo = [NSObject alloc] init`

it actually looks like this in the arc code (simplified):

`
foo = [NSObject alloc];
[foo retain];
[foo init];
[foo release];
....
[foo release];
`

In the normal case `foo` stays the same throughout the init so the consume/release doesn't happen. Think of the non-arc case though where your init returns something different for self

`- (instancetype)init {
     [self release];
     return [gSingletonInstance retain];
}`

We are basically doing the same thing with the mock. We need to release self because we are not returning self as a value. We need to retain the value because ARC is expecting it to be retained so it can release it.